Makefile

```
@@ -213,6 +243,61 @@ $(argocd_sentinel):
 	kubectl -n argocd patch cm argocd-cmd-params-cm -p '{"data": { "server.insecure": "true" } }'
 	kubectl -n argocd patch cm argocd-cm -p '{"data": { "timeout.reconciliation": "30s" } }'
 	kubectl -n argocd rollout restart deployment argocd-server
+	if $(vcluster); then \
+		$(MAKE) argocd-vcluster-auth ; \
+	fi
 	@echo -e "***\n*** Installed argocd in http://argocd.127.0.0.1.nip.io:8088\n***"
 	@echo -e "***\n*** credentials: admin:admin\n***"
 	touch $@
+
+.PHONY: argocd-vcluster-auth
+argocd-vcluster-auth: export KUBECONFIG = $(KIND_KUBECONFIG) ## Re-create argocd authentication for the vcluster, in case it breaks
+argocd-vcluster-auth: vcluster-setup
+argocd-vcluster-auth: vcluster=true
+argocd-vcluster-auth:
+	# The usualy kubeconfig export doesn't work here for some reason...
+	export KUBECONFIG=$(KIND_KUBECONFIG) ; \
+	$(vcluster_bin) connect controlplane --namespace vcluster; \
+	kubectl create serviceaccount argocd; \
+	kubectl create clusterrolebinding argocd_admin --clusterrole=cluster-admin --serviceaccount=default:argocd ; \
+	kubectl apply -f argocd/service-account-secret.yaml ; \
+	sleep 1 ; \
+	export token=$$(kubectl get secret argocd-token -oyaml | yq '.data.token' | base64 -d) ; \
+	$(vcluster_bin) disconnect; \
+	kubectl delete -f argocd/controlplanesecret.yaml ; \
+	cat argocd/controlplanesecret.yaml | yq '.stringData.config = "{ \"bearerToken\":\""+ strenv(token) +"\", \"tlsClientConfig\": { \"insecure\": true }}"' | kubectl apply -f -
+
+.PHONY: install-vcluster-bin
+install-vcluster-bin: $(vcluster_bin)
+
+$(vcluster_bin): export GOOS = $(shell go env GOOS)
+$(vcluster_bin): export GOARCH = $(shell go env GOARCH)
+$(vcluster_bin): export GOBIN = $(go_bin)
+$(vcluster_bin): | $(go_bin)
+	if $(vcluster); then \
+		go install github.com/loft-sh/vcluster/cmd/vclusterctl@latest; \
+	fi
+
+
+.PHONY: vcluster-setup
+vcluster-setup: export KUBECONFIG = $(KIND_KUBECONFIG)
+vcluster-setup: install-vcluster-bin
+	if $(vcluster); then \
+		$(vcluster_bin) create controlplane --namespace vcluster --connect=false -f vclusterconfig/values.yaml || true; \
+	fi
+
+.PHONY: vcluster-in-cluster-kubeconfig
+vcluster-in-cluster-kubeconfig: export KUBECONFIG = $(KIND_KUBECONFIG) ## Prints out a kubeconfig for use within the main cluster
+vcluster-in-cluster-kubeconfig:
+	@export KUBECONFIG=$(KIND_KUBECONFIG) ; \
+	$(vcluster_bin) connect controlplane --namespace vcluster --print | yq '.clusters[0].cluster.server = "https://controlplane.vcluster"'
+
+.PHONY: vcluster-local-cluster-kubeconfig
+vcluster-local-cluster-kubeconfig: export KUBECONFIG = $(KIND_KUBECONFIG) ## Prints out a kubeconfig for use on the local machine
+vcluster-local-cluster-kubeconfig:
+	@export KUBECONFIG=$(KIND_KUBECONFIG) ; \
+	$(vcluster_bin) connect controlplane --namespace vcluster --print | yq
+
+.PHONY: vcluster-clean
+vcluster-clean: ## If you break Crossplane hard enough just remove the whole vcluster
+	$(vcluster_bin) rm controlplane
```

```suggestion
	$(vcluster_bin) rm controlplane || true
```

this way we can run `make clean` few times, helpful, as sometimes kind can't get deleted after first run, at least on my laptop